native/src/widget/text_input.rs

```diff
@@ -1041,3 +841,182 @@ mod platform {
         }
     }
 }
+
+mod cursor {
+    use crate::widget::text_input::Value;
+    use iced_core::Point;
+    use std::time::{Duration, SystemTime};
+
+    /// Even the compiler bullies me for not writing documentation
+    #[derive(Debug, Copy, Clone)]
+    pub struct Cursor {
+        start: usize,
+        end: usize,
```

Comment:
I liked the idea of an enum to represent the cursor state. We should be to create a `cursor::State` enum and maybe expose it through a `Cursor::state` method. I think this should simplify the `Cursor` API a bit and improve/remove some of its methods.

Reply:
So more or less like the first try (Cursor::Index and Cursor::Selection)? Because that sometimes felt a little bit awkward to work with and ended up with more code to use it (always used with match).
Because of that I just ditched it and always use start & end positions and is_selection to check if they are both on the same position.

So i don't know about the simpler API, but you're right about methods needing improvements (->cleanup :P) and that's needed even without your state changes.

But you're the more experienced so I guess I should give it a try

Reply:
> So more or less like the first try (Cursor::Index and Cursor::Selection)?

Yes, but nested here without exposing it directly.

> Because that sometimes felt a little bit awkward to work with and ended up with more code to use it (always used with match).

It may be more verbose, but it made things more explicit. It doesn't make sense to have a `start` and `end` when you do not have an active selection.